engine/engine_impl.go

```diff
@@ -205,51 +205,64 @@ func (k *Kubernetes) Run(ctx context.Context, specv runtime.Spec, stepv runtime.
 	return k.waitForTerminated(ctx, spec, step)
 }
 
-func (k *Kubernetes) waitFor(ctx context.Context, spec *Spec, conditionFunc func(e watch.Event) (bool, error)) error {
+func (k *Kubernetes) waitFor(ctx context.Context, spec *Spec, conditionFunc func(pod *v1.Pod) (bool, error)) error {
 	label := fmt.Sprintf("io.drone.name=%s", spec.PodSpec.Name)
 	lw := &cache.ListWatch{
 		ListFunc: func(options metav1.ListOptions) (k8sruntime.Object, error) {
-			return k.client.CoreV1().Pods(spec.PodSpec.Namespace).List(metav1.ListOptions{
-				LabelSelector: label,
-			})
+			options.LabelSelector = label
+
+			return k.client.CoreV1().Pods(spec.PodSpec.Namespace).List(options)
 		},
 		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			return k.client.CoreV1().Pods(spec.PodSpec.Namespace).Watch(metav1.ListOptions{
-				LabelSelector: label,
-			})
+			options.LabelSelector = label
+
+			return k.client.CoreV1().Pods(spec.PodSpec.Namespace).Watch(options)
 		},
 	}
 
 	preconditionFunc := func(store cache.Store) (bool, error) {
-		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: spec.PodSpec.Namespace, Name: spec.PodSpec.Name})
+		obj, exists, err := store.Get(&metav1.ObjectMeta{Namespace: spec.PodSpec.Namespace, Name: spec.PodSpec.Name})
 		if err != nil {
 			return true, err
 		}
 		if !exists {
-			return true, err
+			return true, fmt.Errorf("pod doesn't exist")
 		}
-		return false, nil
-	}
 
-	_, err := watchtools.UntilWithSync(ctx, lw, &v1.Pod{}, preconditionFunc, conditionFunc)
-	return err
-}
+		pod, ok := obj.(*v1.Pod)
+		if !ok {
+			return true, fmt.Errorf("unexpected object type: %v", obj)
+		}
 
-func (k *Kubernetes) waitForReady(ctx context.Context, spec *Spec, step *Step) error {
-	return k.waitFor(ctx, spec, func(e watch.Event) (bool, error) {
+		return conditionFunc(pod)
```

Comment:
Why are we calling the `conditionFunc` in the `preconditionFunc`?

Reply:
that was my attempt to skip `Watch` altogether if `List` already returns the pod in the state we expect